Tests/NIOTests/SyscallAbstractionLayer.swift

```diff
@@ -0,0 +1,665 @@
+        return try self.withUnsafeFileDescriptor { fd in
+            let buffers = iovecs.map { iovec -> ByteBuffer in
+                var buffer = ByteBufferAllocator().buffer(capacity: iovec.iov_len)
+                buffer.writeBytes(UnsafeRawBufferPointer(start: iovec.iov_base, count: iovec.iov_len))
+                return buffer
+            }
+
+            try self.userToKernel.waitForEmptyAndSet(.writev(fd, buffers))
+            let ret = try self.waitForKernelReturn()
+            if case .returnIOResultInt(let result) = ret {
+                return result
+            } else {
+                throw UnexpectedKernelReturn(ret)
+            }
+        }
+    }
+
+    override func close() throws {
+        let fd = try self.takeDescriptorOwnership()
+
+        try self.userToKernel.waitForEmptyAndSet(.close(fd))
+        let ret = try self.waitForKernelReturn()
+        if case .returnVoid = ret {
+            return
+        } else {
+            throw UnexpectedKernelReturn(ret)
+        }
+    }
+}
+
+extension HookedSelector {
+    func assertSyscallAndReturn(_ result: KernelToUser,
+                                file: StaticString = #file,
+                                line: UInt = #line,
+                                matcher: (UserToKernel) throws -> Bool) throws {
+        let syscall = try self.userToKernel.takeValue()
+        if try matcher(syscall) {
+            try self.kernelToUser.waitForEmptyAndSet(result)
+        } else {
+            XCTFail("unexpected syscall \(syscall)", file: file, line: line)
+            throw UnexpectedSyscall(syscall)
+        }
+    }
+
+    func assertWakeup(file: StaticString = #file, line: UInt = #line) throws {
+        SAL.printIfDebug("\(#function)")
+        try self.assertSyscallAndReturn(.returnSelectorEvent(nil), file: file, line: line) { syscall in
+            if case .whenReady(.block) = syscall {
+                return true
+            } else {
+                return false
+            }
+        }
+        try self.wakeups.takeValue()
+    }
+
+}
+
+extension EventLoop {
+    internal func runSAL<T>(syscallAssertions: () throws -> Void = {},
+                            file: StaticString = #file,
+                            line: UInt = #line,
+                            _ body: @escaping () throws -> T) throws -> T {
+        let hookedSelector = ((self as! SelectableEventLoop)._selector as! HookedSelector)
+        let box = LockedBox<Result<T, Error>>()
+        self.execute {
+            do {
+                try box.waitForEmptyAndSet(.init(catching: body))
+            } catch {
+                box.value = .failure(error)
+            }
+        }
+        try hookedSelector.assertWakeup(file: file, line: line)
+        try syscallAssertions()
+        return try box.takeValue().get()
+    }
+}
+
+extension EventLoopFuture {
+    internal func teleportWait() throws -> Value {
```

Comment:
Why the 'teleport' part of the name, because we're jumping over to the main thread?

Reply:
renamed `salWait` and added a comment.